### PR TITLE
Update `retina` imports to better handle parsing

### PR DESCRIPTION
**Pull request by jmartin-tech**
When parsing Retina XML exports REXML lib in ruby is injecting with space not found inside tags. 
- [x] Adjust for whitespace issue in a backwards compatible way. 
- [x] Add `context` tag parsing to support adding service details when supplied in the XML.
  - Looks about right: `<context>TCP:443 ([redacted]), SHA256[=][redacted], Serial[=][redacted]</context>`
- [x] Add detection for `Retina` files when `Xml Declaration` is present a beginning of the file. 
  - Indeed, the file starts with `<?xml version="1.0" encoding="utf-8"?>` now

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `db_import <PATH_TO_RETINA_XML>`
- [x] **Verify** a sample file imports hosts.

### CODE REVIEW

File: lib/rex/parser/retina_xml.rb [hunk 1]
```diff
@@ -27,6 +27,7 @@ def tag_start(name, attributes)
   end
 
   def text(str)
+    return if str.strip.empty?()
```
Comment (wvu-r7) [inline lib/rex/parser/retina_xml.rb]: Remove parens.

File: lib/msf/core/db_manager/import/retina.rb [hunk 2]
```diff
@@ -74,6 +74,13 @@ def import_retina_xml(args={}, &block)
             :task      => args[:task]
         }
 
+        if vuln['port'] && vuln['proto']
+          vuln_info.merge!({
+               :port => vuln['port'],
```
Comment (wvu-r7) [inline lib/msf/core/db_manager/import/retina.rb]: Two-space indents.

Comment (wvu-r7) [reply lib/msf/core/db_manager/import/retina.rb]: Nevermind. This whole file has messed-up formatting.

File: lib/rex/parser/retina_xml.rb (hunk 1 above)
Comment (wvu-r7) [reply lib/rex/parser/retina_xml.rb]: I'll handle it.

**Comment by wvu-r7** [comment]
I've changed your PR description to have checkboxes so I can verify functionality as I review.

File: lib/rex/parser/retina_xml.rb [hunk 3]
```diff
@@ -59,6 +60,8 @@ def text(str)
       @audit['cce'] = str
     when :in_date
       @audit['data'] = str
+    when :in_context
+      @audit['proto'], @audit['port'] = str.split(/\s+/).first.split(':')
```
Comment (wvu-r7) [inline lib/rex/parser/retina_xml.rb]: For my notes:

```
[1] pry(#<Msf::Framework>)> 'TCP:443 ([redacted]), SHA256[=][redacted], Serial[=][redacted]'.split(/\s+/).first.split(':')
=> ["TCP", "443"]
[2] pry(#<Msf::Framework>)>
```

File: lib/rex/parser/retina_xml.rb [hunk 4]
```diff
@@ -35,7 +36,7 @@ def text(str)
     when :in_netbiosname
       @host["netbios"] = str
     when :in_mac
-      @host["mac"] = str
+      @host["mac"] = str.split(/\s+/).first
```
Comment (wvu-r7) [inline lib/rex/parser/retina_xml.rb]: For my notes:

```
[1] pry(#<Msf::Framework>)> '00:50:56:XX:XX:XX (VMware, Inc.)'.split(/\s+/).first
=> "00:50:56:XX:XX:XX"
[2] pry(#<Msf::Framework>)>
```

File: lib/rex/parser/retina_xml.rb (hunk 1 above)
Comment (wvu-r7) [reply lib/rex/parser/retina_xml.rb]: For my notes:

```
<mac></mac>
```

**Comment by wvu-r7** [comment]
New/fixed behavior:

```
msf5 > db_import ~/Downloads/[redacted]
[*] Importing 'Retina XML' data
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Successfully imported /Users/wvu/Downloads/[redacted]
msf5 >
```

**Comment by wvu-r7** [comment]
Independent fixing/testing to reproduce the issue:

```
msf5 > db_import ~/Downloads/[redacted]
[*] Importing 'IP360 XML v3' data
[-] Failed to import /Users/wvu/Downloads/[redacted]: The nCircle IP360 ASPL file is not present.
    Download ASPL from nCircle VNE | Administer | Support | Resources, unzip it, and import it first
msf5 > git stash pop
[*] exec: git stash pop

On branch master
Your branch is behind 'upstream/master' by 6 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   lib/msf/core/db_manager/import.rb
	modified:   lib/rex/parser/retina_xml.rb

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (d5c6149be15975f4ff64a59a6f9c754eaf9aa5b6)
msf5 > git diff
[*] exec: git diff

diff --git a/lib/msf/core/db_manager/import.rb b/lib/msf/core/db_manager/import.rb
index 9ae2724fd5..252febd480 100644
--- a/lib/msf/core/db_manager/import.rb
+++ b/lib/msf/core/db_manager/import.rb
@@ -407,6 +407,9 @@ module Msf::DBManager::Import
         when /ReportInfo/
           @import_filedata[:type] = "Foundstone"
           return :foundstone_xml
+        when /scanJob/
+          @import_filedata[:type] = "Retina XML"
+          return :retina_xml
         when /ScanGroup/
           @import_filedata[:type] = "Acunetix"
           return :acunetix_xml
diff --git a/lib/rex/parser/retina_xml.rb b/lib/rex/parser/retina_xml.rb
index 16517be6d3..c1abe4c017 100644
--- a/lib/rex/parser/retina_xml.rb
+++ b/lib/rex/parser/retina_xml.rb
@@ -27,6 +27,8 @@ class RetinaXMLStreamParser
   end

   def text(str)
+    return if str.blank?
+
     case @state
     when :in_ip
       @host["address"] = str
@@ -35,7 +37,7 @@ class RetinaXMLStreamParser
     when :in_netbiosname
       @host["netbios"] = str
     when :in_mac
-      @host["mac"] = str
+      @host["mac"] = str.split.first
     when :in_os
       @host["os"] = str
     when :in_rthid
msf5 > reload_lib -a
[*] Reloading /rapid7/metasploit-framework/lib/msf/core/db_manager/import.rb
[*] Reloading /rapid7/metasploit-framework/lib/rex/parser/retina_xml.rb
msf5 > db_import ~/Downloads/[redacted]
[*] Importing 'Retina XML' data
[-]
[-] Warning: The Retina XML format does not associate vulnerabilities with the
[-] specific service on which they were found.
[-] This makes it impossible to correlate exploits to discovered vulnerabilities
[-] in a reliable fashion.
[-]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Importing host [redacted]
[*] Successfully imported /Users/wvu/Downloads/[redacted]
[-] Please note that there was one warning
msf5 > vulns -p 445

Vulnerabilities
===============

Timestamp  Host  Name  References
---------  ----  ----  ----------

msf5 >
```

Good job on the service info. I've removed the warning thusly.

File: lib/rex/parser/retina_xml.rb (hunk 4 above)
Comment (wvu-r7) [reply lib/rex/parser/retina_xml.rb]: Beug.

File: lib/rex/parser/retina_xml.rb [hunk 5]
```diff
@@ -27,6 +26,8 @@ def tag_start(name, attributes)
   end
 
   def text(str)
+    return if str.to_s.strip.empty?
```
Comment (wvu-r7) [inline lib/rex/parser/retina_xml.rb]: Beug.

**Comment by wvu-r7** [comment]
# Release Notes

This updates the Retina XML importer with the current XML format and also enhances the importer with the ability to cross-reference services to vulnerabilities.

**Comment by pbarry-r7** [comment]
Thanks, @jmartin-r7 and @wvu-r7!  👏

**Pull request merged**